### ISSUE TITLE
Unable to deploy function using .NET Core SDK 5.0.102 on Big Sur 11.1

### ISSUE THREAD

**Issue by velocitysystems** [aws/aws-extensions-for-dotnet-cli#155]
## Description
When attempting to deploy a serverless ASP.NET Core Lambda function using `dotnet lambda deploy-serverless`, the CLI tool fails with an MSBuild error and the `dotnet` application crashes.

`Could not load file or assembly 'System.Net.Primitives, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. A device attached to the system is not functioning.`

This occurs when using the .NET Core SDK 5.0.102 on macOS Big Sur 11.1.
**Note: Downgrading to .NET Core SDK 3.1.405 resolves the issue.**

## Reproduction Steps
1. Create a new blank serverless ASP.NET Core Lambda function.
2. Attempt to deploy the function using `dotnet lambda deploy-serverless` using Terminal.
3. The error below is thrown in the output.

### Logs
```
... publish: MSBUILD : error MSB1025: An internal failure occurred while running MSBuild.
... publish: Microsoft.Build.BackEnd.NodeFailedToLaunchException: Could not load file or assembly 'System.Net.Primitives, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. A device attached to the system is not functioning.
... publish:  (0x8007001F)
... publish:  ---> System.IO.FileLoadException: Could not load file or assembly 'System.Net.Primitives, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. A device attached to the system is not functioning.
... publish:  (0x8007001F)
... publish: File name: 'System.Net.Primitives, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'
... publish:    at System.Diagnostics.Process.OpenStream(Int32 fd, FileAccess access)
... publish:    at System.Diagnostics.Process.StartCore(ProcessStartInfo startInfo)
... publish:    at System.Diagnostics.Process.Start()
... publish:    at System.Diagnostics.Process.Start(ProcessStartInfo startInfo)
... publish:    at Microsoft.Build.BackEnd.NodeProviderOutOfProcBase.LaunchNode(String msbuildLocation, String commandLineArgs)
... publish:    --- End of inner exception stack trace ---
... publish:    at Microsoft.Build.BackEnd.NodeProviderOutOfProcBase.LaunchNode(String msbuildLocation, String commandLineArgs)
... publish:    at Microsoft.Build.BackEnd.NodeProviderOutOfProcBase.GetNode(String msbuildLocation, String commandLineArgs, Int32 nodeId, INodePacketFactory factory, Handshake hostHandshake, NodeContextTerminateDelegate terminateNode)
... publish:    at Microsoft.Build.BackEnd.NodeProviderOutOfProc.CreateNode(Int32 nodeId, INodePacketFactory factory, NodeConfiguration configuration)
... publish:    at Microsoft.Build.BackEnd.NodeManager.AttemptCreateNode(INodeProvider nodeProvider, NodeConfiguration nodeConfiguration)
... publish:    at Microsoft.Build.BackEnd.NodeManager.CreateNode(NodeConfiguration configuration, NodeAffinity nodeAffinity)
... publish:    at Microsoft.Build.Execution.BuildManager.PerformSchedulingActions(IEnumerable`1 responses)
... publish:    at Microsoft.Build.Execution.BuildManager.HandleNewRequest(Int32 node, BuildRequestBlocker blocker)
... publish:    at Microsoft.Build.Execution.BuildManager.ProcessPacket(Int32 node, INodePacket packet)
... publish:    at Microsoft.Build.Execution.BuildManager.<>c__DisplayClass70_0.<Microsoft.Build.BackEnd.INodePacketHandler.PacketReceived>b__0()
... publish:    at Microsoft.Build.Execution.BuildManager.ProcessWorkQueue(Action action)
... publish: --- End of stack trace from previous location ---
... publish:    at Microsoft.Build.Execution.BuildManager.EndBuild()
... publish:    at Microsoft.Build.CommandLine.MSBuildApp.BuildProject(String projectFile, String[] targets, String toolsVersion, Dictionary`2 globalProperties, Dictionary`2 restoreProperties, ILogger[] loggers, LoggerVerbosity verbosity, DistributedLoggerRecord[] distributedLoggerRecords, Int32 cpuCount, Boolean enableNodeReuse, TextWriter preprocessWriter, TextWriter targetsWriter, Boolean detailedSummary, ISet`1 warningsAsErrors, ISet`1 warningsAsMessages, Boolean enableRestore, ProfilerLogger profilerLogger, Boolean enableProfiler, Boolean interactive, Boolean isolateProjects, Boolean graphBuild, Boolean lowPriority, String[] inputResultsCaches, String outputResultsCache)
```

### Environment
* Build Version: Amazon Lambda Tools for .NET Core applications (5.0.0)
* OS Info: macOS Big Sur 11.1
* Build Environment: Visual Studio | VSCode + Terminal `dotnet`
* Targeted .NET Platform: .NET Core 3.1 LTS

## Resolution
No resolution, only workaround is to remove SDK 5.0.102 from the system.

- [ ] :wave: I can/would-like-to implement a fix for this problem myself

---

This is a :bug: bug-report

**Comment by ashishdhingra** [comment]
Hi @velocitysystems,

Good morning.

Could you please confirm if this reproducible on macOS Catalina 10.15.7 or Windows system? From high level, it appears to be .NET Core 5.0 SDK bug, the error occurs during MSBuild process.

Thanks,
Ashish

**Comment by velocitysystems** [comment]
Thanks @ashishdhingra.

**I can confirm this does not occur on Windows 10 Version 1909 using .NET SDK Version: 5.0.102.**

I agree this appears to be an issue with the .NET Core 5.0 SDK on macOS. Unfortunately, I don't have a macOS Catalina 10.15.7 environment to test in sorry.

**Comment by ashishdhingra** [comment]
> Thanks @ashishdhingra.
> 
> **I can confirm this does not occur on Windows 10 Version 1909 using .NET SDK Version: 5.0.102.**
> 
> I agree this appears to be an issue with the .NET Core 5.0 SDK on macOS. Unfortunately, I don't have a macOS Catalina 10.15.7 environment to test in sorry.

I will try to reproduce this on macOS Catalina and let you know the results. I'm not sure if this is AWS Lambda Tools issue though since the error occurs during MSBuild process.

**Comment by ashishdhingra** [comment]
Hi @velocitysystems,

Somehow I'm unable to reproduce this on macOS Catalina 10.15.7. Here are the steps:

1. Verified dotnet --version. (Returned `5.0.102`)
2. Installed AWS Lambda Templates using command `dotnet new -i "Amazon.Lambda.Templates::*"`.
3. Created ASP.NET Serverless Web app using command `dotnet new serverless.AspNetCoreWebApp --name AspNetServerLess`
4. Created S3 bucket `testaspnetwebapp` in regions `us-east-2`.
4. Executed command `dotnet lambda deploy-serverless --stack-name testaspnetwebapp --s3-bucket testaspnetwebapp --region us-east-2`:
```
Amazon Lambda Tools for .NET Core applications (5.0.0)
Project Home: https://github.com/aws/aws-extensions-for-dotnet-cli, https://github.com/aws/aws-lambda-dotnet
	
Processing CloudFormation resource AspNetCoreApi
Processing CloudFormation resource AspNetCoreFunction
Initiate packaging of . for resource AspNetCoreFunction
Executing publish command
... invoking 'dotnet publish', working folder '/Users/<<username>>/dev/AspNetServerLess/src/AspNetServerLess/./bin/Release/netcoreapp3.1/publish'
... dotnet publish --output "/Users/<<username>>/dev/AspNetServerLess/src/AspNetServerLess/./bin/Release/netcoreapp3.1/publish" --configuration "Release" --framework "netcoreapp3.1" /p:GenerateRuntimeConfigurationFiles=true --runtime linux-x64 --self-contained false 
... publish: Microsoft (R) Build Engine version 16.8.3+39993bd9d for .NET
... publish: Copyright (C) Microsoft Corporation. All rights reserved.
... publish:   Determining projects to restore...
... publish:   Restored /Users/ashdhin/dev/AspNetServerLess/src/AspNetServerLess/AspNetServerLess.csproj (in 228 ms).
... publish:   AspNetServerLess -> /Users/<<username>>/dev/AspNetServerLess/src/AspNetServerLess/bin/Release/netcoreapp3.1/linux-x64/AspNetServerLess.dll
... publish:   AspNetServerLess -> /Users/<<username>>/dev/AspNetServerLess/src/AspNetServerLess/bin/Release/netcoreapp3.1/linux-x64/AspNetServerLess.Views.dll
... publish:   AspNetServerLess -> /Users/<<username>>/dev/AspNetServerLess/src/AspNetServerLess/bin/Release/netcoreapp3.1/publish/
Changed permissions on published file (chmod +rx Amazon.Lambda.APIGatewayEvents.dll).
Changed permissions on published file (chmod +rx AspNetServerLess.Views.dll).
Changed permissions on published file (chmod +rx AspNetServerLess.deps.json).
Changed permissions on published file (chmod +rx AspNetServerLess).
Changed permissions on published file (chmod +rx Amazon.Lambda.Logging.AspNetCore.dll).
Changed permissions on published file (chmod +rx AspNetServerLess.Views.pdb).
Changed permissions on published file (chmod +rx web.config).
Changed permissions on published file (chmod +rx appsettings.json).
Changed permissions on published file (chmod +rx Amazon.Lambda.ApplicationLoadBalancerEvents.dll).
Changed permissions on published file (chmod +rx AspNetServerLess.runtimeconfig.json).
Changed permissions on published file (chmod +rx Amazon.Lambda.Serialization.SystemTextJson.dll).
Changed permissions on published file (chmod +rx aws-lambda-tools-defaults.json).
Changed permissions on published file (chmod +rx AspNetServerLess.pdb).
Changed permissions on published file (chmod +rx Amazon.Lambda.Core.dll).
Changed permissions on published file (chmod +rx bundleconfig.json).
Changed permissions on published file (chmod +rx AspNetServerLess.dll).
Changed permissions on published file (chmod +rx appsettings.Development.json).
Changed permissions on published file (chmod +rx Amazon.Lambda.AspNetCoreServer.dll).
Zipping publish folder /Users/<<username>>/dev/AspNetServerLess/src/AspNetServerLess/./bin/Release/netcoreapp3.1/publish to /var/folders/5v/klr4lq9j0rd95h4xtmydvgrwqf_g0k/T/AspNetCoreFunction-CodeUri-Or-ImageUri-637463064921454550.zip
... zipping:   adding: Amazon.Lambda.APIGatewayEvents.dll (deflated 57%)
... zipping:   adding: AspNetServerLess.Views.dll (deflated 73%)
... zipping:   adding: AspNetServerLess.deps.json (deflated 91%)
... zipping:   adding: AspNetServerLess (deflated 61%)
... zipping:   adding: Amazon.Lambda.Logging.AspNetCore.dll (deflated 50%)
... zipping:   adding: AspNetServerLess.Views.pdb (deflated 51%)
... zipping:   adding: web.config (deflated 40%)
... zipping:   adding: appsettings.json (deflated 25%)
... zipping:   adding: Amazon.Lambda.ApplicationLoadBalancerEvents.dll (deflated 49%)
... zipping:   adding: AspNetServerLess.runtimeconfig.json (deflated 36%)
... zipping:   adding: Amazon.Lambda.Serialization.SystemTextJson.dll (deflated 48%)
... zipping:   adding: aws-lambda-tools-defaults.json (deflated 46%)
... zipping:   adding: AspNetServerLess.pdb (deflated 48%)
... zipping:   adding: Amazon.Lambda.Core.dll (deflated 47%)
... zipping:   adding: bundleconfig.json (deflated 55%)
... zipping:   adding: AspNetServerLess.dll (deflated 57%)
... zipping:   adding: appsettings.Development.json (deflated 13%)
... zipping:   adding: Amazon.Lambda.AspNetCoreServer.dll (deflated 58%)
... zipping:   adding: wwwroot/favicon.ico (deflated 71%)
... zipping:   adding: wwwroot/css/site.min.css (deflated 32%)
... zipping:   adding: wwwroot/css/site.css (deflated 46%)
... zipping:   adding: wwwroot/js/site.min.js (stored 0%)
... zipping:   adding: wwwroot/js/site.js (stored 0%)
... zipping:   adding: wwwroot/lib/bootstrap/LICENSE (deflated 41%)
... zipping:   adding: wwwroot/lib/bootstrap/.bower.json (deflated 51%)
... zipping:   adding: wwwroot/lib/jquery/.bower.json (deflated 49%)
... zipping:   adding: wwwroot/lib/jquery/LICENSE.txt (deflated 44%)
... zipping:   adding: wwwroot/lib/jquery-validation-unobtrusive/jquery.validate.unobtrusive.js (deflated 76%)
... zipping:   adding: wwwroot/lib/jquery-validation-unobtrusive/jquery.validate.unobtrusive.min.js (deflated 63%)
... zipping:   adding: wwwroot/lib/jquery-validation-unobtrusive/.bower.json (deflated 59%)
... zipping:   adding: wwwroot/lib/jquery-validation/LICENSE.md (deflated 42%)
... zipping:   adding: wwwroot/lib/jquery-validation/.bower.json (deflated 54%)
... zipping:   adding: wwwroot/lib/jquery/dist/jquery.min.map (deflated 60%)
... zipping:   adding: wwwroot/lib/jquery/dist/jquery.js (deflated 70%)
... zipping:   adding: wwwroot/lib/jquery-validation/dist/additional-methods.js (deflated 72%)
... zipping:   adding: wwwroot/lib/jquery-validation/dist/jquery.validate.js (deflated 73%)
... zipping:   adding: wwwroot/lib/bootstrap/dist/css/bootstrap.css (deflated 86%)
... zipping:   adding: wwwroot/lib/bootstrap/dist/css/bootstrap.css.map (deflated 81%)
... zipping:   adding: wwwroot/lib/bootstrap/dist/css/bootstrap.min.css.map (deflated 81%)
... zipping:   adding: wwwroot/lib/bootstrap/dist/css/bootstrap-theme.css.map (deflated 85%)
... zipping:   adding: wwwroot/lib/bootstrap/dist/css/bootstrap-theme.min.css.map (deflated 80%)
... zipping:   adding: wwwroot/lib/bootstrap/dist/css/bootstrap-theme.css (deflated 89%)
... zipping:   adding: wwwroot/lib/bootstrap/dist/js/npm.js (deflated 62%)
... zipping:   adding: wwwroot/lib/bootstrap/dist/js/bootstrap.js (deflated 79%)
... zipping:   adding: wwwroot/lib/bootstrap/dist/fonts/glyphicons-halflings-regular.woff (deflated 1%)
... zipping:   adding: wwwroot/lib/bootstrap/dist/fonts/glyphicons-halflings-regular.eot (deflated 0%)
... zipping:   adding: wwwroot/lib/bootstrap/dist/fonts/glyphicons-halflings-regular.woff2 (deflated 0%)
... zipping:   adding: wwwroot/lib/bootstrap/dist/fonts/glyphicons-halflings-regular.ttf (deflated 48%)
... zipping:   adding: wwwroot/lib/bootstrap/dist/fonts/glyphicons-halflings-regular.svg (deflated 75%)
Created publish archive (/var/folders/5v/klr4lq9j0rd95h4xtmydvgrwqf_g0k/T/AspNetCoreFunction-CodeUri-Or-ImageUri-637463064921454550.zip).
Lambda project successfully packaged: /var/folders/5v/klr4lq9j0rd95h4xtmydvgrwqf_g0k/T/AspNetCoreFunction-CodeUri-Or-ImageUri-637463064921454550.zip
Uploading to S3. (Bucket: testaspnetwebapp Key: AspNetServerLess/AspNetCoreFunction-CodeUri-Or-ImageUri-637463064921454550-637463064956821960.zip)
... Progress: 15%
... Progress: 30%
... Progress: 46%
... Progress: 61%
... Progress: 77%
... Progress: 92%
... Progress: 100%
Uploading to S3. (Bucket: testaspnetwebapp Key: AspNetServerLess/testaspnetwebapp-serverless-637463065008851450.template)
... Progress: 100%
Found existing stack: False
CloudFormation change set created
... Waiting for change set to be reviewed
Created CloudFormation stack testaspnetwebapp
   
Timestamp            Logical Resource Id                      Status                                   
-------------------- ---------------------------------------- ---------------------------------------- 
1/15/21 11:21 AM     testaspnetwebapp                         CREATE_IN_PROGRESS                      
1/15/21 11:21 AM     AspNetCoreFunctionRole                   CREATE_IN_PROGRESS                      
1/15/21 11:21 AM     AspNetCoreFunctionRole                   CREATE_IN_PROGRESS                      
1/15/21 11:22 AM     AspNetCoreFunctionRole                   CREATE_COMPLETE                         
1/15/21 11:22 AM     AspNetCoreFunction                       CREATE_IN_PROGRESS                      
1/15/21 11:22 AM     AspNetCoreFunction                       CREATE_IN_PROGRESS                      
1/15/21 11:22 AM     AspNetCoreFunction                       CREATE_COMPLETE                         
1/15/21 11:22 AM     AspNetCoreApi                            CREATE_IN_PROGRESS                      
1/15/21 11:22 AM     AspNetCoreApi                            CREATE_IN_PROGRESS                      
1/15/21 11:22 AM     AspNetCoreApi                            CREATE_COMPLETE                         
1/15/21 11:22 AM     AspNetCoreFunctionProxyResourcePermissionProd CREATE_IN_PROGRESS                      
1/15/21 11:22 AM     AspNetCoreApiDeployment22eefd9c68        CREATE_IN_PROGRESS                      
1/15/21 11:22 AM     AspNetCoreFunctionRootResourcePermissionProd CREATE_IN_PROGRESS                      
1/15/21 11:22 AM     AspNetCoreFunctionProxyResourcePermissionProd CREATE_IN_PROGRESS                      
1/15/21 11:22 AM     AspNetCoreFunctionRootResourcePermissionProd CREATE_IN_PROGRESS                      
1/15/21 11:22 AM     AspNetCoreApiDeployment22eefd9c68        CREATE_IN_PROGRESS                      
1/15/21 11:22 AM     AspNetCoreApiDeployment22eefd9c68        CREATE_COMPLETE                         
1/15/21 11:22 AM     AspNetCoreApiProdStage                   CREATE_IN_PROGRESS                      
1/15/21 11:22 AM     AspNetCoreApiProdStage                   CREATE_IN_PROGRESS                      
1/15/21 11:22 AM     AspNetCoreApiProdStage                   CREATE_COMPLETE                         
1/15/21 11:22 AM     AspNetCoreFunctionRootResourcePermissionProd CREATE_COMPLETE                         
1/15/21 11:22 AM     AspNetCoreFunctionProxyResourcePermissionProd CREATE_COMPLETE                         
1/15/21 11:22 AM     testaspnetwebapp                         CREATE_COMPLETE                         
Stack finished updating with status: CREATE_COMPLETE
   
Output Name                    Value                                             
------------------------------ --------------------------------------------------
ApiURL                         https://7a1ejst2w6.execute-api.us-east-2.amazonaws.com/Prod/
``` 

NOTE: For `Lambda ASP.NET Core Web API (.NET 5 Container Image)`, you need to use command `dotnet new serverless.image.AspNetCoreWebAPI --name AspNetServerLessImage` from .NET CLI.

**Comment by ashishdhingra** [comment]
Hi @velocitysystems, 

Just to clarify, you are targeting .NET Core 3.1 runtime, but using AWS Lambda tools 5.0 version. Please confirm.

Thanks,
Ashish

**Comment by velocitysystems** [comment]
Thanks @ashishdhingra, yes that is correct.
Removing the .NET 5 SDK the deploy succeeds.

**Comment by ashishdhingra** [comment]
> Thanks @ashishdhingra, yes that is correct.
> Removing the .NET 5 SDK the deploy succeeds.

Thanks. Also try running the MSBuild command when .NET Core SDK 5.0.102 is installed and see if still gives error. Although you have confirmed that you are targeting .NET Core 3.1, kindly also check the .csproj file to make sure that `TargetFramework` is `netcoreapp3.1`. Here's what my .csproj file looks like:

```XML
<Project Sdk="Microsoft.NET.Sdk.Web">
  <PropertyGroup>
    <TargetFramework>netcoreapp3.1</TargetFramework>
    <GenerateRuntimeConfigurationFiles>true</GenerateRuntimeConfigurationFiles>
    <AWSProjectType>Lambda</AWSProjectType>
    <!-- This property makes the build directory similar to a publish directory and helps the AWS .NET Lambda Mock Test Tool find project dependencies. -->
    <CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Amazon.Lambda.AspNetCoreServer" Version="5.3.0" />
  </ItemGroup>
  <ItemGroup>
    <Folder Include="wwwroot\images\" />
  </ItemGroup>
</Project>
```

 Most probably this is a BigSur issue (just trying to rule out all possibilities) and you might need to open a case with Microsoft.

Thanks,
Ashish

**Comment by velocitysystems** [comment]
Thanks @ashishdhingra. I can confirm my .csproj file includes the following:

```
<Project Sdk="Microsoft.NET.Sdk.Web">
  <PropertyGroup>
    <TargetFramework>netcoreapp3.1</TargetFramework>
    <GenerateRuntimeConfigurationFiles>true</GenerateRuntimeConfigurationFiles>
    <AWSProjectType>Lambda</AWSProjectType>
  </PropertyGroup>
```

I dig some further testing and think I may have isolated the problem. I have two _two_ Lambda .NET Core web projects. One of them builds and deploys without any errors, the other throws the error as reported.

The one that throws the error, has these additional lines in the .csproj:

```
  <Target Name="DebugScript" Condition="'$(Configuration)'=='Debug'" BeforeTargets="AfterBuild">
    <Exec Command="gulp dev" />
  </Target>
  <Target Name="PrepublishScript" BeforeTargets="PrepareForPublish">
    <Exec Command="gulp clean" />
  </Target>
  <Target Name="PostpublishScript" AfterTargets="AfterPublish">
    <Exec Command="gulp build-deploy" />
  </Target>
```

This allows me to run gulp commands to build SASS, package/minify assets as part of the build process.
It seems when trying to execute these CLI commands (which execute using Node), the error as reported is received.

In summary, it appears this is an MSBuild bug with .NET Core 5 and possibly macOS Big Sur.
I'm happy to close this issue @ashishdhingra and I will report it directly to Microsoft. Thanks so much for your support!

**Comment by smarts** [comment]
For anyone else who encounters this issue and uses ESET:
I encountered this same problem and it was due to the `Enable Real-time file system protection` checkbox being checked in ESET.
